backend/app/notify/email.go

```diff
@@ -0,0 +1,162 @@
+package notify
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	log "github.com/go-pkgz/lgr"
+	"github.com/pkg/errors"
+	"gopkg.in/gomail.v2"
+)
+
+// EmailParams contain settings for email set up
+type EmailParams struct {
+	Server    string
+	Port      int
+	Username  string
+	Password  string
+	KeepAlive time.Duration
+}
+
+// Email implements notify.Destination for email
+type Email struct {
+	server    string
+	port      int
+	username  string
+	password  string
+	keepAlive time.Duration
+	sendChan  chan *gomail.Message
+	errChan   chan error
+}
+
+const emailConnectionKeepAlive = 30 * time.Second
+
+//NewEmail makes email object for notifications and run sending daemon
+func NewEmail(ctx context.Context, params EmailParams) (*Email, error) {
+
+	res := Email{
+		server:    params.Server,
+		port:      params.Port,
+		username:  params.Username,
+		password:  params.Password,
+		keepAlive: params.KeepAlive,
+		sendChan:  make(chan *gomail.Message),
+		errChan:   make(chan error),
+	}
+
+	if res.keepAlive == 0 {
+		res.keepAlive = emailConnectionKeepAlive
+	}
+
+	log.Printf(
+		"[DEBUG] create new email notifier for server %s with user %s, keepalive=%s",
+		res.server, res.username, res.keepAlive)
+
+	// Test connection before starting a daemon.
+	tmpConn, err := gomail.NewDialer(res.server, res.port, res.username, res.password).Dial()
+	if err != nil {
+		return &res, errors.Errorf(
+			"[WARN] error establishing test connecting to '%s':%d with username '%s': %s",
+			res.server, res.port, res.username, err)
+	}
+	if err = tmpConn.Close(); err != nil {
+		return &res, errors.Errorf(
+			"[WARN] error closing test connection to %s:%d: %s",
+			res.server, res.port, err)
+	}
+
+	// Activate server goroutine.
+	go res.activate()
+	// Close the channel to stop the mail daemon
+	go func() { <-ctx.Done(); close(res.sendChan) }()
+
+	return &res, nil
+}
+
+// Send email from request to address in settings
+func (e *Email) Send(ctx context.Context, req request) error {
+	log.Printf("[DEBUG] send notification via %s, comment id %s", e, req.comment.ID)
+
+	messageBody := prepareBody(req)
+
+	// Create message.
+	m := gomail.NewMessage()
+	// TODO: figure out where "from" and "to" addresses come from
+	//m.SetHeader("From", "no-reply@example.com")
+	//m.SetAddressHeader("To", req.Address, req.Name)
+	m.SetHeader("Subject", fmt.Sprintf("New comment for \"%s\"", req.comment.PostTitle))
+	m.SetBody("text/html", messageBody)
+
+	// Wait for ability to send message and return error from error channel after sending it.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case e.sendChan <- m:
+		return <-e.errChan
```

Comment:
hmm, here you effectively enforce synced send. If your goal to send is syncroniosely why even bother with channel?

Reply:
To keep the connection alive and close it after a period without activity, we need a goroutine with a timer inside: I don't know another good way to close the connection by time in the event-based system.
If we have a goroutine which closes the connection, it makes sense to establish a connection and send emails from there as well.
From what I know communication between goroutines achieved with channels.
The only way around I come with so far is establishing a connection from Send (maintaining lock around it), and closing it right away if no one else waiting for the lock. It's still synced, but more complex in my opinion.
Is having multiple "event receivers" and single "event executer" in different goroutines and communicating through channels is a problem? And if so what is a better approach for keeping the stateful connection?